File: desktop/electron/src/telemetry/device-identity.ts
import { execFileSync } from 'node:child_process'
import { createHash } from 'node:crypto'
import { closeSync, openSync, readSync } from 'node:fs'
import { win32 } from 'node:path'

import type { DesktopPlatform } from './contracts.js'

const DEVICE_ID_DOMAIN = 'opensquilla.telemetry.device.v1\0'
const MACHINE_ID_RE = /^[a-f0-9]{32}$/
const DEVICE_ID_RE = /^[a-f0-9]{64}$/

/** Hash only an OS machine identifier; profile paths and network addresses are never inputs. */
export function deriveDeviceId(platform: DesktopPlatform, machineId: string): string | null {
  const normalized = machineId.trim().replaceAll('-', '').toLowerCase()
  if (
    !['macos', 'windows', 'linux'].includes(platform)
    || !MACHINE_ID_RE.test(normalized)
    || /^0+$/.test(normalized)
    || /^f+$/.test(normalized)
  ) return null
  return createHash('sha256').update(`${DEVICE_ID_DOMAIN}${platform}\0${normalized}`, 'utf8')
    .digest('hex')
}

interface DeviceIdentityOptions {
  platform?: NodeJS.Platform
  windowsDirectory?: string
  readText?: (path: string) => string
  run?: (command: string, args: readonly string[]) => string
}

function readMachineIdFile(path: string): string {
  const descriptor = openSync(path, 'r')
  try {
    const buffer = Buffer.alloc(128)
    const length = readSync(descriptor, buffer, 0, buffer.length, 0)
    return length < buffer.length ? buffer.subarray(0, length).toString('utf8') : ''
  } finally {
    closeSync(descriptor)
  }
}

/** Lazily read once per process, only when an already-consented producer requests an ID. */
export function createDeviceIdProvider(options: DeviceIdentityOptions = {}): () => string | null {
  let cached: string | null | undefined
  const platform = options.platform ?? process.platform
  const readText = options.readText ?? readMachineIdFile
  const run = options.run ?? ((command: string, args: readonly string[]) => execFileSync(
    command, [...args], { encoding: 'utf8', timeout: 1_500, maxBuffer: 64 * 1024, windowsHide: true,
      stdio: ['ignore', 'pipe', 'pipe'] },
  ))
  return () => {
    if (cached !== undefined) return cached
    cached = null
    try {
      if (platform === 'darwin') {
        const output = run('/usr/sbin/ioreg', ['-rd1', '-c', 'IOPlatformExpertDevice'])
        const match = /"IOPlatformUUID"\s*=\s*"([0-9A-Fa-f-]+)"/.exec(output)
        cached = match ? deriveDeviceId('macos', match[1]) : null
      } else if (platform === 'win32') {
        const directory = options.windowsDirectory ?? process.env.SystemRoot ?? 'C:\\Windows'
        if (!/^[A-Za-z]:[\\/]/.test(directory)) return null
        const output = run(win32.join(directory, 'System32', 'reg.exe'), [
          'query', 'HKLM\\SOFTWARE\\Microsoft\\Cryptography', '/v', 'MachineGuid', '/reg:64',
        ])
        const match = /^\s*MachineGuid\s+REG_SZ\s+([0-9A-Fa-f-]+)\s*$/im.exec(output)
        cached = match ? deriveDeviceId('windows', match[1]) : null
      } else if (platform === 'linux') {
        for (const path of ['/etc/machine-id', '/var/lib/dbus/machine-id']) {
          try { cached = deriveDeviceId('linux', readText(path)) } catch { /* try next OS file */ }
          if (cached !== null) break
        }
      }
    } catch {
      // Unavailable or malformed OS identities never fall back to random profile identities.
    }
    return cached
  }
}

export const getDeviceId = createDeviceIdProvider()

export function deviceIdentityFields(provider: () => string | null): { device_id?: string } {
  try {
    const value = provider()
    return typeof value === 'string' && DEVICE_ID_RE.test(value) ? { device_id: value } : {}
  } catch {
    return {}
  }
}

File: src/opensquilla/telemetry/device_identity.py
"""Derive an application-specific device identifier without exporting OS identifiers.

Callers must check upload policy before resolving identity. Profile directories,
user accounts and network interfaces are deliberately not identity inputs.
An unavailable OS identity leaves device-based metrics unattributed.
"""

from __future__ import annotations

import hashlib
import re
import subprocess
import sys
from functools import lru_cache
from pathlib import Path

_DOMAIN = "opensquilla.telemetry.device.v1\0"
_MACHINE_ID_RE = re.compile(r"[0-9a-f]{32}\Z")
_IO_PLATFORM_UUID_RE = re.compile(r'"IOPlatformUUID"\s*=\s*"([0-9A-Fa-f-]+)"')
_LOOKUP_TIMEOUT_SECONDS = 1.0


def derive_device_id(platform: str, machine_id: str) -> str | None:
    """Return a purpose-separated SHA-256 token from a valid OS machine ID."""

    if platform not in {"macos", "windows", "linux"}:
        return None
    normalized = machine_id.strip().replace("-", "").lower()
    if (
        not _MACHINE_ID_RE.fullmatch(normalized)
        or normalized == "0" * 32
        or normalized == "f" * 32
    ):
        return None
    return hashlib.sha256(f"{_DOMAIN}{platform}\0{normalized}".encode()).hexdigest()


@lru_cache(maxsize=1)
def get_device_id() -> str | None:
    """Resolve once per process; never substitute a random or network identity."""

    try:
        if sys.platform == "darwin":
            raw = _macos_machine_id()
            return derive_device_id("macos", raw) if raw is not None else None
        if sys.platform == "win32":
            raw = _windows_machine_id()
            return derive_device_id("windows", raw) if raw is not None else None
        if sys.platform.startswith("linux"):
            for path in ("/etc/machine-id", "/var/lib/dbus/machine-id"):
                raw = _read_machine_id_file(Path(path))
                device_id = derive_device_id("linux", raw) if raw is not None else None
                if device_id is not None:
                    return device_id
    except Exception:
        # Never propagate an OS error that could contain the original value.
        return None
    return None


def _macos_machine_id() -> str | None:
    result = subprocess.run(
        ["/usr/sbin/ioreg", "-rd1", "-c", "IOPlatformExpertDevice"],
        stdin=subprocess.DEVNULL,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
        check=True,
        text=True,
        timeout=_LOOKUP_TIMEOUT_SECONDS,
    )
    if len(result.stdout) > 256 * 1024:
        return None
    match = _IO_PLATFORM_UUID_RE.search(result.stdout)
    return match.group(1) if match is not None else None


def _windows_machine_id() -> str | None:
    if sys.platform != "win32":
        return None
    import winreg

    with winreg.OpenKey(
        winreg.HKEY_LOCAL_MACHINE,
        r"SOFTWARE\Microsoft\Cryptography",
        0,
        winreg.KEY_READ | winreg.KEY_WOW64_64KEY,
    ) as key:
        value, value_type = winreg.QueryValueEx(key, "MachineGuid")
    return value if value_type == winreg.REG_SZ and isinstance(value, str) else None


def _read_machine_id_file(path: Path) -> str | None:
    try:
        with path.open("r", encoding="ascii") as stream:
            value = stream.read(129)
        return value if len(value) <= 128 else None
    except (OSError, UnicodeError):
        return None


__all__ = ["derive_device_id", "get_device_id"]
